File: miner.pl
#!/usr/bin/env perl
use strict;
use warnings;
use IPC::Open2;
use Symbol qw(gensym);

my $deep     = 0;
my $clean    = 1;
my $all      = 0;
my @corpora;

while (@ARGV) {
    my $arg = shift @ARGV;

    if ($arg eq '-d' || $arg eq '--deep') {
        $deep = 1;
    }
    elsif ($arg eq '--no-clean') {
        $clean = 0;
    }
    elsif ($arg eq '-a' || $arg eq '--all') {
        $all = 1;
    }
    elsif ($arg eq '-h' || $arg eq '--help') {
        usage();
        exit 0;
    }
    elsif ($arg =~ /^-/) {
        die "Unknown option: $arg\n";
    }
    else {
        push @corpora, $arg;
    }
}

if (!@corpora) {
    @corpora = ("$ENV{HOME}/Chinese Text Analysis/");

    if ($deep) {
        push @corpora,
            "$ENV{HOME}/src/TurnBasedGameData/",
            "$ENV{HOME}/src/AnimeGameData/";
    }
}

my @rg_base = (
    'rg',
    '--no-filename',
    '-N',
    '-uu',
    @corpora,
    '-g', '!old/',
    '-g', '!Anki_dump/',
);

if ($deep) {
    push @rg_base, '--max-filesize', '100M';
}
else {
    push @rg_base, '-ttxt', '-tmd', '--max-filesize', '5M';
}

while (my $line = <STDIN>) {
    $line =~ s/[\r\n]+$//;
    next if $line =~ /^\s*$/;

    # Input format: TARGET whitespace ANCHOR TEXT
    # The anchor may contain spaces.
    my ($target, $anchor) = split /\s+/, $line, 2;
    next unless defined $target && defined $anchor && length $target && length $anchor;

    my $target_pattern = variant_regex($target);
    my @anchor_variants = variants($anchor);

    open my $rg, '-|', @rg_base, '-e', $target_pattern
        or die "Could not run rg: $!\n";

    while (my $hit = <$rg>) {
        next unless contains_any($hit, @anchor_variants);

        $hit = clean_line($hit) if $clean;
        print "$target\t$hit\n";

        last unless $all;
    }

    close $rg;
}

sub variants {
    my ($text) = @_;

    my $simp = opencc('t2s.json', $text);
    my $trad = opencc('s2t.json', $text);

    my %seen;
    return grep { !$seen{$_}++ } ($text, $simp, $trad);
}

sub variant_regex {
    my ($text) = @_;
    return join '|', map { quotemeta($_) } variants($text);
}

sub contains_any {
    my ($line, @needles) = @_;

    for my $needle (@needles) {
        return 1 if index($line, $needle) >= 0;
    }

    return 0;
}

sub clean_line {
    my ($line) = @_;

    $line =~ s/[\r\n]+$//;
    $line =~ s/^\[.*?\]\s*//;     # [146782006]
    $line =~ s/<[^>]+>//g;         # Unity/HTML-style tags
    $line =~ s/^[[:space:]]+//;
    $line =~ s/[[:space:]]+$//;

    return $line;
}

sub opencc {
    my ($config, $text) = @_;

    my $err = gensym;
    my $pid = open2(my $out, my $in, 'opencc', '-c', $config);

    print {$in} $text;
    close $in;

    local $/;
    my $converted = <$out> // '';
    close $out;
    waitpid($pid, 0);

    $converted =~ s/[\r\n]+$//;
    return $converted;
}

sub usage {
    print <<'USAGE';
Usage:
  printf 'TARGET ANCHOR\n' | miner.pl [options] [corpus ...]

Examples:
  printf '期颐 少年\n' | miner.pl -d
  printf '桡骨 右臂\n' | miner.pl ~/src/TurnBasedGameData/
  cat targets.txt | miner.pl --deep --all

Options:
  -d, --deep      Search Chinese Text Analysis plus HSR/Genshin dumps.
  -a, --all       Print all matching lines instead of first match only.
  --no-clean      Do not strip bracket IDs or formatting tags.
  -h, --help      Show this help.
USAGE
}

File: miner.py
#!/usr/bin/env python3

import argparse
import html
import re
import subprocess
import sys
from pathlib import Path

HOME = Path.home()

NORMAL_PATHS = [
    HOME / "Chinese Text Analysis",
]

DEEP_PATHS = [
    HOME / "Chinese Text Analysis",
    HOME / "src" / "TurnBasedGameData",
    HOME / "src" / "AnimeGameData",
]


def run_opencc(text: str, config: str) -> str:
    result = subprocess.run(
        ["opencc", "-c", config],
        input=text,
        text=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        check=True,
    )
    return result.stdout.strip()


def variants(text: str) -> list[str]:
    simp = run_opencc(text, "t2s.json")
    trad = run_opencc(text, "s2t.json")

    seen = []
    for item in [text, simp, trad]:
        if item and item not in seen:
            seen.append(item)

    return seen


def clean_line(line: str) -> str:
    line = line.rstrip("\r\n")

    # Strip leading bracketed IDs, e.g. [146782006]
    line = re.sub(r"^\[.*?\]\s*", "", line)

    # Strip Unity/HTML-ish tags, e.g. <color=#dbc291ff>, </color>
    line = re.sub(r"<[^>]+>", "", line)

    # Decode entities if any show up
    line = html.unescape(line)

    return line.strip()


def rg_search(targets: list[str], paths: list[Path], deep: bool) -> list[str]:
    cmd = [
        "rg",
        "--fixed-strings",
        "--no-filename",
        "-N",
    ]

    for target in targets:
        cmd.extend(["-e", target])

    if deep:
        cmd.extend(["--max-filesize", "100M"])
    else:
        cmd.extend(["-ttxt", "-tmd", "--max-filesize", "5M"])

    cmd.extend([
        "-g", "!old/",
        "-g", "!Anki_dump/",
    ])

    cmd.extend(str(path) for path in paths)

    result = subprocess.run(
        cmd,
        text=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )

    if result.returncode not in (0, 1):
        print(result.stderr, file=sys.stderr, end="")
        sys.exit(result.returncode)

    return result.stdout.splitlines()


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Mine Chinese sentence examples by target + anchor text."
    )
    parser.add_argument(
        "-d", "--deep",
        action="store_true",
        help="Search HSR/Genshin dump folders too.",
    )
    parser.add_argument(
        "--all",
        action="store_true",
        help="Print all matching lines instead of the first match.",
    )
    parser.add_argument(
        "--no-clean",
        action="store_true",
        help="Do not strip IDs or tags from output.",
    )
    parser.add_argument(
        "paths",
        nargs="*",
        help="Optional corpus paths. Overrides default/deep paths.",
    )

    args = parser.parse_args()

    if args.paths:
        paths = [Path(p).expanduser() for p in args.paths]
    else:
        paths = DEEP_PATHS if args.deep else NORMAL_PATHS

    for raw in sys.stdin:
        raw = raw.strip()
        if not raw:
            continue

        parts = raw.split(maxsplit=1)
        if len(parts) < 2:
            print(f"Skipping line without target + anchor: {raw}", file=sys.stderr)
            continue

        target, anchor = parts
        target_variants = variants(target)
        anchor_variants = variants(anchor)

        hits = rg_search(target_variants, paths, args.deep)

        for hit in hits:
            if not any(anchor_variant in hit for anchor_variant in anchor_variants):
                continue

            output = hit if args.no_clean else clean_line(hit)
            print(f"{target}\t{output}")

            if not args.all:
                break


if __name__ == "__main__":
    main()
